File: html/js/pbclass/PbFunctionDefinition.js
// PbFunctionDefinition.js — port of Java PbFunctionDefinition.java
import { PbFunctionFlag } from './PbFunctionFlag.js';

export class PbFunctionDefinition {
    constructor() {
        /** @type {import('./PbObject.js').PbObject|null} */
        this.object      = null;
        /** @type {number} */
        this.index       = 0;
        /** @type {number} */
        this.globalIndex = 0;
        /** @type {number} */
        this.refIndex    = 0;
        /** @type {number} */
        this.eventCode   = 0;
        /** @type {number} */
        this.flagValue   = 0;
        /** @type {import('./PbType.js').PbType|null} */
        this.returnType  = null;
        /** @type {string} */
        this.name        = '';
        /** @type {import('./PbFunctionParam.js').PbFunctionParam[]} */
        this.params      = [];
        /** @type {string|null} */
        this.library     = null;
        /** @type {string|null} */
        this.alias       = null;
        /** @type {import('./PbType.js').PbType|null} */
        this.throwsType  = null;
    }

    isEvent()    { return PbFunctionFlag.hasFlag(this.flagValue, PbFunctionFlag.IsEvent); }
    isExternal() { return PbFunctionFlag.hasFlag(this.flagValue, PbFunctionFlag.IsExternal); }

    toString() {
        let text = '';
        if (!this.isEvent()) {
            if (PbFunctionFlag.hasFlag(this.flagValue, PbFunctionFlag.IsPrivate)) {
                text += 'private ';
            } else if (!PbFunctionFlag.hasFlag(this.flagValue, PbFunctionFlag.IsProtected)) {
                text += 'public ';
            } else {
                text += 'protected ';
            }
            text += (this.returnType.index !== 0) ? 'function ' : 'subroutine ';
        } else {
            text += 'event ';
        }
        text += this.returnType.name + ' ';
        const paramList = this.params.map(p => p.toString()).join(',');
        text += `${this.name}(${paramList})`;
        if (this.throwsType !== null)  text += ` throws ${this.throwsType.name}`;
        if (this.library   !== null)   text += ` library "${this.library}" alias for "${this.alias}"`;
        return text;
    }
}

File: java/src/main/java/com/pbdviewer/utils/pbclass/PbFunctionDefinition.java
package com.pbdviewer.utils.pbclass;

import java.util.Arrays;
import java.util.stream.Collectors;

public class PbFunctionDefinition {

    private PbObject object;
    private int index;
    private int globalIndex;
    private int refIndex;
    private int eventCode;
    private int flagValue;
    private PbType returnType;
    private String name;
    private PbFunctionParam[] params;
    private String library;
    private String alias;
    private PbType throwsType;

    public boolean isEvent() {
        return PbFunctionFlag.hasFlag(flagValue, PbFunctionFlag.IsEvent);
    }

    public boolean isExternal() {
        return PbFunctionFlag.hasFlag(flagValue, PbFunctionFlag.IsExternal);
    }

    @Override
    public String toString() {
        String text = "";
        if (!isEvent()) {
            if (PbFunctionFlag.hasFlag(flagValue, PbFunctionFlag.IsPrivate)) {
                text += "private ";
            } else if (!PbFunctionFlag.hasFlag(flagValue, PbFunctionFlag.IsProtected)) {
                text += "public ";
            } else {
                text += "protected ";
            }
            text += returnType.getIndex() != 0 ? "function " : "subroutine ";
        } else {
            text += "event ";
        }
        text += returnType.getName() + " ";
        String paramList = Arrays.stream(params).map(Object::toString).collect(Collectors.joining(","));
        text += String.format("%s(%s)", name, paramList);
        if (throwsType != null) text += String.format(" throws %s", throwsType.getName());
        if (library != null) text += String.format(" library \"%s\" alias for \"%s\"", library, alias);
        return text;
    }

    public PbObject getObject() {
        return object;
    }

    public void setObject(PbObject object) {
        this.object = object;
    }

    public int getIndex() {
        return index;
    }

    public void setIndex(int index) {
        this.index = index;
    }

    public int getGlobalIndex() {
        return globalIndex;
    }

    public void setGlobalIndex(int globalIndex) {
        this.globalIndex = globalIndex;
    }

    public int getRefIndex() {
        return refIndex;
    }

    public void setRefIndex(int refIndex) {
        this.refIndex = refIndex;
    }

    public int getEventCode() {
        return eventCode;
    }

    public void setEventCode(int eventCode) {
        this.eventCode = eventCode;
    }

    public int getFlag() {
        return flagValue;
    }

    public void setFlag(int flagValue) {
        this.flagValue = flagValue;
    }

    public PbType getReturnType() {
        return returnType;
    }

    public void setReturnType(PbType returnType) {
        this.returnType = returnType;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public PbFunctionParam[] getParams() {
        return params;
    }

    public void setParams(PbFunctionParam[] params) {
        this.params = params;
    }

    public String getLibrary() {
        return library;
    }

    public void setLibrary(String library) {
        this.library = library;
    }

    public String getAlias() {
        return alias;
    }

    public void setAlias(String alias) {
        this.alias = alias;
    }

    public PbType getThrowsType() {
        return throwsType;
    }

    public void setThrowsType(PbType throwsType) {
        this.throwsType = throwsType;
    }
}
